databases/core.py

```diff
@@ -344,6 +344,9 @@ def __init__(
         self._connection_callable = connection_callable
         self._force_rollback = force_rollback
         self._extra_options = kwargs
+        self._transaction_context: ContextVar[TransactionBackend | None] = ContextVar(
+            "transaction_context"
+        )
```

Comment:
fyi https://docs.python.org/3/library/contextvars.html#contextvars.ContextVar

> Important: [Context](https://docs.python.org/3/library/contextvars.html#contextvars.Context) Variables should be created at the top module level and never in closures. Context objects hold strong references to context variables which prevents context variables from being properly garbage collected.

Reply:
Yeah, I read that but it didn't click until coming back to review this. I added some comments to the code since the design around these contextvars is pretty opaque and foreign for a lot of people when reading it.

As the docs says, the reason that context variables _should_ be created at module level is that they automatically register with the current `contextvars.Context`, which creates a strong (aka counted) reference from the context to the new context variable. Since the Context keeps a reference to the context variable, won't be garbage collected until the Context is destroyed. Many async programs will use one long-lived asyncio.Task (containing a single Context), and even with my changes Databases will still be leaking a ContextVar with every call to `async with db.transaction()`, or the lower level `transaction = await db.transaction()`. In my view, decorator-usage (eg for a FastAPI route) will not leak these ContextVars, since only one is created per `databases.Transaction` instance, which is used (and re-ured) as the wrapping decorator. 

The problem with module level decorators for Transactions is that two transactions could happen in the same task, and these need to remain as independent instances. Take this example: 

```py
import asyncio
import databases
db = databases.Database(...)

async def nested_transactions():
    async with db.transaction():
        db.execute(...)
        async with db.transaction():
            db.execute(...)

async def main() -> None:
    # Non-concurrent example, but with multiple Tasks
    await asyncio.create_task(nested_transactions)

asyncio.run(main())
```

The above script needs to maintain two separate `databases.interfaces.TransactionBackend`  instances, one for the inner transaction, and one for the outer. We of course also need to keep the code safe for concurrent execution, so plain globals are out (due to race conditions just like if we were using threads). If these was a single module level context variable to store the transaction's backend, the nested transaction would overwrite the outer's reference. That's why I stuck with the instance level contextvar.

Reply:
Although, with writing that out, I wonder if a module level context variable that holds a task-local dictionary that maps `databases.Transaction` instances to `databases.interfaces.TransactionBackend` instances could work. We'd get one module level context variable, but still be able to store references to multiple transaction backend objects per task...

Reply:
Also, Databases currently has a non-global context variable to store the `databases.Database`'s connection instance. In similar fashion, it seems that we couldn't use a module level context variable for that either, since there could be usage like: 

```python
import asyncio
import databases

db1 = databases.Database(...)
db2 = databases.Database(...)

async def copy_something():
    something = db.execute(...)
    db.execute(process(something)

asyncio.run(main)
```

Obviously, in this example, we need two independent connection objects, one for each instance. But, perhaps the `ContextVar[dict[Database, Connection]]` type could work here too...

Reply:
> If these was a single module level context variable to store the transaction's backend, the nested transaction would overwrite the outer's reference. That's why I stuck with the instance level contextvar.

If you reset the variable with the token the outer context's value would be restored when the inner context exits, right?

Reply:
Yes - it seems so. Using a token to reset a value brings back the value that was current before the contextvar.set(...) call was made - regardless of the order the tokens are used. The old value is simply stored on the token.

```python
>>> import contextvars
>>> var = contextvars.ContextVar('var', default={'id': 'default'})
>>> token0 = var.set({'id': 1})
>>> token1 = var.set({'id': 2})
>>> token2 = var.set({'id': 3})
>>> token3 = var.set({'id': 4})
>>> var.get()
{'id': 4}
>>> var.reset(token1)
>>> var.get()
{'id': 1}
>>> var.reset(token3)
>>> var.get()
{'id': 3}
>>> var.reset(token2)
>>> var.get()
{'id': 2}
>>> var.reset(token0)
>>> var.get()
{'id': 'default'}
>>> token3.old_value
{'id': 3}
```

Reply:
Resolving: new implementation avoids using context variables